Improve multi-image support for STM32 UART and timer drivers

duda-patryk opened the pull request:
Hi!

These changes add UART and timer reset using RCC. This is needed in multi-image environment to make sure that the devices are initialized correctly and won't cause system crash or hang.

Best regards,
Patryk

duda-patryk committed:
drivers: counter: Reset timer using RCC before initialization

In multi-image environment, after jump to the image we can have running
timer with interrupts enabled. If interrupt is triggered, the asserts
in the driver can cause a crash.

This patch also adds 'resets' property for all timer nodes.

Signed-off-by: Patryk Duda <pdk@semihalf.com>

duda-patryk committed:
drivers: serial: Reset UART using RCC before initialization

In multi-image environment, after jump to the image we can have UART in
unexpected state. Reset UART to default state to make sure that UART is
initialized properly and won't cause system to crash or hang.

Signed-off-by: Patryk Duda <pdk@semihalf.com>

semihalf-duda-patryk commented:
Friendly ping.

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -1615,7 +1615,8 @@ static int uart_stm32_init(const struct device *dev)
 	}
 
-	LL_USART_Disable(config->usart);
+	/* Reset UART to default state using RCC */
+	LL_USART_DeInit(config->usart);
```
These are 2 different things. Removing `LL_USART_Disable` may be valid. But this should be done in a follow up commit.

semihalf-duda-patryk replied:
Removing `LL_USART_Disable` is valid because after `LL_USART_DeInit` all UART registers have default values. Moved removing `LL_USART_Disable` to separate commit.

erwango commented on drivers/counter/counter_ll_stm32_timer.c:
```diff
@@ -470,6 +470,9 @@ static int counter_stm32_init_timer(const struct device *dev)
 	}
 	data->freq = tim_clk / (cfg->prescaler + 1U);
 
+	/* Reset timer to default state using RCC */
+	LL_TIM_DeInit(timer);
```
If we're to add device reset, which will eventually end up in most drivers, I think the clean way is to add include/zephyr/drivers/reset.h support.
I acknowledge this is a significantly larger change though...

semihalf-duda-patryk replied:
Could you confirm that I understand it correctly:
- I need to add reset driver for STM32 and modify DTS for all STM32 devices appropriately.
- In the driver I will use RCC to reset requested module.
- Use reset api in UART and timer (and modify their DTS, I guess).

I agree that this approach is much better, cleaner in a long term. If I have any questions I'll let you know.

erwango replied:
Your understanding is correct. This is a significantly larger change (but relatively simple).

semihalf-duda-patryk replied:
I added reset drivers and used reset API in UART and timer drivers. It was **significant** amount of work. I wouldn't say it was simple, but definitely worth it.

erwango reviewed (changes requested):
I'd prefer we directly aim for the reset API implementation

duda-patryk committed:
drivers: reset: Introduce STM32 reset controller

This driver exposes STM32 RCC reset functionality through reset API.

Information about RCC register offset and bit is encoded just like GD32.
The first 5 least significant bits contains register bit number.
Next 12 bits are used to keep RCC register offset. Remaining bits are
unused.

Signed-off-by: Patryk Duda <pdk@semihalf.com>

duda-patryk committed:
dts: arm: st: u5: Add reset controller node

Reset controller node is necessary to enable support for resetting
peripherials using RCC.

This patch also includes RCC reset registers offsets used by STM32_RESET
macro.

Signed-off-by: Patryk Duda <pdk@semihalf.com>

duda-patryk committed:
dts: arm: st: g0: Add reset controller node

Reset controller node is necessary to enable support for resetting
peripherials using RCC.

This patch also includes RCC reset registers offsets used by STM32_RESET
macro.

Signed-off-by: Patryk Duda <pdk@semihalf.com>

duda-patryk committed:
dts: arm: st: g4/l4/l5: Add reset controller node

Reset controller node is necessary to enable support for resetting
peripherials using RCC.

This patch also includes RCC reset registers offsets used by STM32_RESET
macro.

Signed-off-by: Patryk Duda <pdk@semihalf.com>

duda-patryk committed:
dts: arm: st: h7: Add reset controller node

Reset controller node is necessary to enable support for resetting
peripherials using RCC.

This patch also includes RCC reset registers offsets used by STM32_RESET
macro.

Signed-off-by: Patryk Duda <pdk@semihalf.com>

duda-patryk committed:
dts: arm: st: mp1: Add reset controller node

Reset controller node is necessary to enable support for resetting
peripherials using RCC.

This patch also includes RCC reset registers offsets used by STM32_RESET
macro.

Signed-off-by: Patryk Duda <pdk@semihalf.com>

duda-patryk committed:
dts: arm: st: l0: Add reset controller node

Reset controller node is necessary to enable support for resetting
peripherials using RCC.

This patch also includes RCC reset registers offsets used by STM32_RESET
macro.

Signed-off-by: Patryk Duda <pdk@semihalf.com>

duda-patryk committed:
dts: arm: st: l1: Add reset controller node

Reset controller node is necessary to enable support for resetting
peripherials using RCC.

This patch also includes RCC reset registers offsets used by STM32_RESET
macro.

Signed-off-by: Patryk Duda <pdk@semihalf.com>

duda-patryk committed:
dts: arm: st: wb/wl: Add reset controller node

Reset controller node is necessary to enable support for resetting
peripherials using RCC.

This patch also includes RCC reset registers offsets used by STM32_RESET
macro.

Signed-off-by: Patryk Duda <pdk@semihalf.com>

duda-patryk committed:
dts: arm: st: f0/f1/f3: Add reset controller node

Reset controller node is necessary to enable support for resetting
peripherials using RCC.

This patch also includes RCC reset registers offsets used by STM32_RESET
macro.

Signed-off-by: Patryk Duda <pdk@semihalf.com>

duda-patryk committed:
dts: arm: st: f2/f4/f7: Add reset controller node

Reset controller node is necessary to enable support for resetting
peripherials using RCC.

This patch also includes RCC reset registers offsets used by STM32_RESET
macro.

Signed-off-by: Patryk Duda <pdk@semihalf.com>

semihalf-duda-patryk commented:
Added reset driver for STM32 and STM32MP1. Used reset API in UART and timer drivers.

erwango commented on include/zephyr/dt-bindings/reset/stm32f4_reset.h:
```diff
@@ -0,0 +1,103 @@
+#define STM32_RESET_BUS_APB2 0x24
+
+/* AHB1 */
+#define STM32_RESET_GPIOA    STM32_RESET(AHB1, 0U)
+#define STM32_RESET_GPIOB    STM32_RESET(AHB1, 1U)
+#define STM32_RESET_GPIOC    STM32_RESET(AHB1, 2U)
+#define STM32_RESET_GPIOD    STM32_RESET(AHB1, 3U)
+#define STM32_RESET_GPIOE    STM32_RESET(AHB1, 4U)
+#define STM32_RESET_GPIOF    STM32_RESET(AHB1, 5U)
+#define STM32_RESET_GPIOG    STM32_RESET(AHB1, 6U)
+#define STM32_RESET_GPIOH    STM32_RESET(AHB1, 7U)
+#define STM32_RESET_GPIOI    STM32_RESET(AHB1, 8U)
+#define STM32_RESET_GPIOJ    STM32_RESET(AHB1, 9U)
+#define STM32_RESET_GPIOK    STM32_RESET(AHB1, 10U)
+#define STM32_RESET_CRC	     STM32_RESET(AHB1, 12U)
+#define STM32_RESET_DMA1     STM32_RESET(AHB1, 21U)
+#define STM32_RESET_DMA2     STM32_RESET(AHB1, 22U)
+#define STM32_RESET_DMA2D    STM32_RESET(AHB1, 23U)
+#define STM32_RESET_ETHMAC   STM32_RESET(AHB1, 25U)
+#define STM32_RESET_OTGHS    STM32_RESET(AHB1, 29U)
+#define STM32_RESET_RNG_AHB1 STM32_RESET(AHB1, 31U)
```
Rather than providing this header, it would be easier to provide direct implementation in .dsti files:
```
			gpioa: gpio@40020000 {
				compatible = "st,stm32-gpio";
				gpio-controller;
				#gpio-cells = <2>;
				reg = <0x40020000 0x400>;
				clocks = <&rcc STM32_CLOCK_BUS_AHB1 0x00000001>;
				reset = <&rcc STM32_RESET(AHB1, 0U)>;
			};
```
which is more in line with what is done for clocks property and prevents the need to have SoC specific files (stm32f413_reset.h, stm32f446_reset.h).

semihalf-duda-patryk replied:
I took GD32 approach as example (eg. include/zephyr/dt-bindings/reset/gd32e10x.h). I will follow your suggestion and change it.

semihalf-duda-patryk replied:
Besides of this, does the code look good for you?

erwango replied:
Still looking into it, but this was the most visible issue I saw.
Otherwise the remaining part looks good, but I wanted to warn earlier on that part (I wish I could have warn before you perform the complete device tree completion).
Then, it would be nice to propose a specific PR for this whole development which is much boarder than "Improve multi-image support for STM32 UART and timer drivers"

erwango reviewed (changes requested):
Thanks for this great addition. One request regarding the device tree definition.

semihalf-duda-patryk commented:
I replaced defines in DTS with `STM32_RESET()`. To do that, I used magic power of C preprocessor and sed.

semihalf-duda-patryk commented:
Added missing newline at the end of include/zephyr/dt-bindings/reset/stm32u5_reset.h file.

erwango commented on drivers/reset/reset_stm32mp1.c:
```diff
@@ -0,0 +1,79 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT st_stm32mp1_rcc_rctl
```
Instead of duplicating the whole driver, can we think about a (n optional) property than could be use to differentiate behaviors?

semihalf-duda-patryk replied:
Yes, it's possible. I would change `reset_stm32_line_deassert()` to support STM32MP1 + some define changes in driver. Do you think it would be cleaner? Do you have a name for the property?

PS I found bug in drivers/reset/reset_stm32mp1.c just now, so this is not final version.

erwango replied:
One idea would be to introduce 2 props on all series which would be set-offset/clr-offset with value 5/5 on most series except MP1 where it would be 5/17.

semihalf-duda-patryk replied:
To deassert reset on STM32MP1 you need to set but in clear register (clearing bit in set/clear register does nothing), so we actually need 3 properties. What do you think about it? Can I use default value for set-offset/clr-offset, and add these properties in stm32mp1.dtsi?

erwango replied:
Alternative is to use the following for MP1:
```
         compatible = "st,stm32mp1-rcc-rctl", "st,stm32-rcc-rctl";
```
This way you get both compatibles defined when compiled for MP1 case.
Then you can use `DT_HAS_COMPAT_ENABLED(st_stm32mp1_rcc_rct)` to tweak stm32 reset driver behavior in MP1 case. Similar trick is used in SPI and I2C STM32 drivers.

semihalf-duda-patryk replied:
I added `set-bit-to-deassert` property. It's used in MP1 `rctl` node to indicate that the driver should set bit in clear register to deassert reset.

duda-patryk committed:
drivers: reset: Add support for reset clear register

To deassert reset in STM32MP1 RCC the driver needs to set the bit in
reset clear register.

This patch extends existing implementation to support this type of
register.

Signed-off-by: Patryk Duda <pdk@semihalf.com>

erwango commented on dts/bindings/reset/st,stm32-rcc-rctl.yaml:
```diff
@@ -27,5 +27,11 @@ properties:
+    set-bit-to-deassert:
+      type: boolean
+      description: |
+        Indicates if the drives should set bit in clear register to
```
```suggestion
        Indicates if the driver should set bit in clear register to
```

erwango commented on dts/arm/st/f0/stm32f0.dtsi:
```diff
@@ -231,6 +231,7 @@
 			compatible = "st,stm32-timers";
 			reg = <0x40012c00 0x400>;
 			clocks = <&rcc STM32_CLOCK_BUS_APB2 0x00000800>;
+			resets = <&rctl STM32_RESET(APB2, 11U)>;
```
dts changes should not be in this commit (drivers: counter: Reset timer using RCC before initialization)

semihalf-duda-patryk replied:
Fixed.

erwango commented on dts/arm/st/f0/stm32f0.dtsi:
```diff
@@ -159,6 +159,7 @@
 			compatible = "st,stm32-usart", "st,stm32-uart";
 			reg = <0x40013800 0x400>;
 			clocks = <&rcc STM32_CLOCK_BUS_APB2 0x00004000>;
+			resets = <&rctl STM32_RESET(APB2, 14U)>;
```
Ditto. dts changes don't belong to this commit

erwango reviewed (changes requested):
wrong button: some changes are actually required

duda-patryk committed:
dts: Introduce 'resets' property for STM32 timer nodes

We are about to add timer reset during driver initialization. First step
is to add 'resets' property, which provides information about reset
register offset and bit.

Signed-off-by: Patryk Duda <pdk@semihalf.com>

duda-patryk committed:
dts: Introduce 'resets' property to STM32 UART nodes

We are about to add UART reset during driver initialization. First step
is to add 'resets' property, which provides information about reset
register offset and bit.

Signed-off-by: Patryk Duda <pdk@semihalf.com>

semihalf-duda-patryk commented:
Split changes in DTS from changes in driver. Fixed typo.

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -1618,7 +1619,13 @@ static int uart_stm32_init(const struct device *dev)
 		return err;
 	}
 
-	LL_USART_Disable(config->usart);
```
This should be kept.
On some series (H7 at least),  device shjould be disabled to set some registers such as baudrate

semihalf-duda-patryk replied:
I think RCC reset changes UE bit (Uart Enable) to 0, but I can keep it to be extra safe.

erwango replied:
ok right. That should be easy to check anyway. But let's keep that for an extra PR if needed.

semihalf-duda-patryk commented:
Restored call to `LL_UART_Disable`
